Bug Fixes

BlueViper85 opened the pull request:
Looking at the bug fixes, I created a fork of the project and coded a few of the fixes myself, at the very least as a means to learn more about the project and improve my on C# skills and in case anyone found any use from it.

-Fixed crash caused by using <seed> or <date> in a folder name for the ROM File (Issue #279
-Fixed the logic for the chests in the Eye bridge room at the end of Turtle Rock to prevent the Lamp from spawning here (Issue #274)
-Added CanLiftLightRocks(have) as a requirement for the BumperCave, since CanAccessNorthwestDarkWorld(have) can evaluate to true without having the ability to lift light rocks (as it should) (Issue #281)
-Updated CanDefeatHera(have) so that it will check if you can light torches if the big key is in the 1F chest, this will help with issues where Moldorm could have the Fire Rod (or possibly the Lamp) which could prevent you from reaching him. (Issue #266)
-Updated the logic for the Big Chest in Swamp Palace to cover a more broad spectrum of big key locations than it originally covered to help ensure it can be opened. (Issue #267)

BlueViper85 committed:
Added <seed> and <date> support for folder paths

Per https://github.com/Dessyreqt/alttprandomizer/issues/279, the Randomizer client would crash if you used <seed> or <date> in the names of the folders on the way to the ROM file.  Added support for this.

BlueViper85 committed:
Add files via upload

-Fixed the logic for the Eye bridge room at the end of Turtle Rock to prevent the Lamp from spawning here (see: https://github.com/Dessyreqt/alttprandomizer/issues/274)

-Added CanLiftLightRocks(have) as a requirement for the BumperCave, since CanAccessNorthwestDarkWorld(have) can evaluate to true without having the ability to lift light rocks (see: https://github.com/Dessyreqt/alttprandomizer/issues/266 and https://github.com/Dessyreqt/alttprandomizer/issues/281)

-Changed CanDefeatHera() so that it will check if you can light torches if the big key is in the 1F chest (see: https://github.com/Dessyreqt/alttprandomizer/issues/276

-Changed the logic for the Big Chest in Swamp Palace so that it checks either that Hookshot is available OR that the Big Key is not in one of the late-dungeon chests so that it checks the chests even if the Hookshot is available, instead of stopping at the Hookshot and sort of hoping that the Big Key doesn't wind up in those chests. (see: https://github.c

BlueViper85 committed:
Fix for Swamp Big Chest Logic

Modified the logic for the Big Chests further than my previous update as my previous change was insufficient.

This new code checks that the Big Key is properly accessible in the first 3/4 of Swamp Palace.  If it's not, then it makes sure you have Hookshot and that the Big Chest does not contain said hookshot.  As long as those two items are true you can get to the remaining chests and will be able to open the Big Chest.

sporchia commented on AlttpRandomizer/Rom/RomLocationsCasual.cs:
```diff
@@ -103,11 +103,12 @@ public void ResetLocations()
+                                || LocationHasItem("[dungeon-D2-B1] Swamp Palace - map room", InventoryItemType.BigKey)
+                                || LocationHasItem("[dungeon-D2-B1] Swamp Palace - push 4 blocks room", InventoryItemType.BigKey)
+                                || LocationHasItem("[dungeon-D2-B1] Swamp Palace - south of hookshot room", InventoryItemType.BigKey))
+                            || (have.Contains(InventoryItemType.Hookshot) && !LocationHasItem("[dungeon-D2-B1] Swamp Palace - big chest",InventoryItemType.Hookshot)))
```
this line seems weird to me

BlueViper85 replied:
This is on the Big Chest logic

First it checks to see if the Big Key has been placed in the chests you can access without the hoookshot.

If it's not in any of those chests, it checks to see if you have access to the hookshot. If you don't it returns False telling the Randomizer you can't access the big chest.

If you do have access to the hookshot, it also checks to make sure it's not in the big chest itself.  If it's not, again you can access the last three chests and get the big key if it were in there.

If the big key hasn't been placed at all it already shouldn't think you can access it anyway.

It's not perfect but it's the best method i've been able to come up with that didn't generate RandomizationExceptions when ingenerated 100 seeds. I want to find a better way, and plan to continue looking at it but at least for now I'm happy with the generation for my own use (wouldn't dare use it to race against others for obvious reasons)

If anything stands out as weird or if you have ideas for improvements I'd love to hear it. Been a great learning process, but tons more to go.

Dessyreqt replied:
Please follow the convention of having each case on a separate line. 

Looks like you have an unnecessary set of parens around those four rooms where you are checking for the big key (since they are followed by another ||.)

BlueViper85 replied:
Thanks, Dessy!  I've corrected the last line so each case is on it's own line, and I removed those extra parentheses

Dessyreqt commented on AlttpRandomizer/Random/Randomizer.cs:
```diff
@@ -75,7 +75,8 @@ public string CreateRom(RandomizerOptions options)
                 if (options.Filename.Contains("\\") && !Directory.Exists(options.Filename.Substring(0, options.Filename.LastIndexOf('\\'))))
                 {
-                    Directory.CreateDirectory(options.Filename.Substring(0, options.Filename.LastIndexOf('\\')));
+                    String filePath = FileName.Fix(options.Filename, string.Format(romLocations.SeedFileString, seed), complexity);
```
Seems like it would be better to use FileName.Fix on the directory before checking to see if it exists instead of after.

BlueViper85 replied:
Of course you are correct, I have moved the creation of the filePath string to occur just before the if statement and updated the if statement to use that instead of the original path.

BlueViper85 committed:
Formatting Updates

Had a logic statement with two conditions on the same line, moved them to the same line to conform to the established standard, and removed an unnecessary set of parentheses.

BlueViper85 committed:
Formatting Updates

Fixed the formatting of the Swamp Palace big chest logic to remove an extra set of parentheses and to split the last check for the Hookshot placement onto two separate lines, to maintain the established standard.

BlueViper85 committed:
ROM File Path

Moved the creation of the filePath string to occur before the initial if statement and updated the if statement accordingly so it will check that the fixed file path exists, where it was previously still relying on the "non-fixed" directory

BlueViper85 committed:
Trinexx Lamp

Updated CanDefeatTurtleRock(have) to include a check for the lamp as well, this will keep the lamp from spanwing at Trinexx's heart piece

BlueViper85 committed:
Magic Upgrade in Spoiler Log

Added the chosen Magic Upgrade to the end of the spoiler log, without specifying its location so it can be looked up if desired.

BlueViper85 committed:
Swamp Palace and Magic Bat Fixes

Removed the additional checks on the Big Chest in swamp palace and restored it to it's original logic from v7
Added a "CanAccessLateSwampPalace" function which does the following:

Checks that you have the hammer
Then it checks for the big key in any of the other chests in Swamp Palace that don't require the big key
If none of them have the big key THEN it checks if you have the hookshot.
If you have the hookshot it then makes sure that it's not in the big chest.

I associated "CanAccessLateSwampPalace(have)" to the requirements for the two chests in the flooded room, and the Hidden Waterfall door room.

I regenerated the seed with this being the ONLY modification to the code from v7 and the hookshot was still in the big chest but the Big Key moved to the Big Key chest.  I created 100 more seeds, all of which created without any Randomization Exceptions and the big key was all over Swamp Palace, including the three late-dungeon chests.  This feels like a far superio

BlueViper85 commented:
I've made a number of changes to this code to address additional issues, in addition to the above I've added the following fixes.  Some of these I've mentioned in other comments, but at this point should probably be specified here as well:

-Moved the FilePath variable to be created before the If statement that checks if the path exists so it can use the corrected path when evaluating if it exists or not.
-I failed to add the Lamp to the check for Trinexx as well, so I added it as a requirement to CanDefeatTurtleRock(have)
-I also added the Light World region to Magic Bat so it will appear in the Spoiler Log properly
-Added a function that finds the Magic Upgrade in the seed, and outputs which magic upgrade was chosen (1/2 or 1/4 magic) and specifies which at the bottom of the spoiler log (without specifying where it is)chosen Magic Upgrade to the bottom of the spoiler log (without specifying where it is).  This and adding the Magic Bat back to the spoiler log address issue #290 
-Updated the CanGetGoldSword logic to confirm that the player can take the Super Bomb to the Pyramid.  This includes a check for the moon pearl for when Crystals can be in the Light World.  I ran a seed through before and after changing the code and it didn't change item placements at all for that seed to ensure no unintended side effects came out of this change.

I also completely re-did how I handled the logic for the Hookshot being in the Big Chest in Swamp Palace.  I wasn't comfortable with it as it was since I couldn't confirm if the Hookshot would still spawn in the Big Chest.

These are the changes I made:
-Removed the additional checks on the Big Chest in swamp palace and restored it to it's original logic from v7
-Added a "CanAccessLateSwampPalace" function which does the following:
-Associated the new "CanAccessLateSwampPalace(have)" to the requirements for the two chests in the flooded room, and the one in the hidden waterfall door room.

This new function performs the following checks:
-Checks that you have the hammer
-Then it checks for the big key in any of the other chests in Swamp Palace that don't require the hookshot.  If they have it, it proceeds. If the big key isn't there, then:
-If none of them have the big key THEN it checks if you have the hookshot
-If you have access to the hookshot it then makes sure that it's not in the big chest.

I took this new function and chest associations and put them into the original v7 code (instead of my fork) and regenerated the original seed (C5532622) with this being the ONLY modification to the code from v7 and the hookshot was still in the big chest but the Big Key moved to the Big Key chest.  I created 100 more seeds, all of which created without any Randomization Exceptions and the big key was all over Swamp Palace, including the three late-dungeon chests.  This should be a more solid fix for Issue #267

BlueViper85 committed:
Updated the "CanGetGoldSword" logic for Casual to include logic that ensures you can take the Super Bomb to the Pyramid
